File: aat-lib/src/main/java/ch/bailu/aat_lib/util/fs/AppDirectory.java
package ch.bailu.aat_lib.util.fs;

import java.io.IOException;
import java.util.Locale;

import ch.bailu.aat_lib.preferences.map.SolidTileCacheDirectory;
import ch.bailu.aat_lib.preferences.system.SolidDataDirectory;
import ch.bailu.foc.Foc;

public class AppDirectory  {

    //////////////////////////////////////////////////////////////////////////////////////
    public static Foc getDataDirectory(SolidDataDirectory sdirectory, String sub) {
        return sdirectory.getValueAsFile().descendant(sub);
    }

    public static final String DIR_AAT_DATA = "aat_data";
    public static final String DIR_TILES = "tiles";
    public static final String DIR_TILES_OSMDROID = "osmdroid/tiles";


    private static final String DIR_LOG = "log";
    private static final String FILE_LOG = "log.gpx";
    public static Foc getLogFile(SolidDataDirectory sdirectory) {
        return getDataDirectory(sdirectory, DIR_LOG).child(FILE_LOG);
    }


    public static final String DIR_OVERLAY = "overlay";
    public static final String DIR_IMPORT = "import";

    public static final String DIR_NOMINATIM = "query/nominatim";
    public static final String DIR_OVERPASS = "query/overpass";
    public static final String DIR_POI = "query/poi";


    public static final String DIR_TEST = "test";





    public static final String DIR_CACHE = "cache";
    public static final String FILE_CACHE_DB="summary.db";



    private static final String DIR_EDIT = "overlay/draft";
    private static final String FILE_DRAFT= "draft.gpx";

    public static Foc getEditorDraft(SolidDataDirectory sdirectory) {
        return getDataDirectory(sdirectory, DIR_EDIT).child(FILE_DRAFT);
    }

    //////////////////////////////////////////////////////////////////////////////////////




    /////////////////////////////////////////////////////////////////////////////////////////////////
    public static Foc getTileFile(String tilePath, SolidTileCacheDirectory sdirectory) {
        return getTileCacheDirectory(sdirectory).descendant(tilePath);
    }


    private static Foc getTileCacheDirectory(SolidTileCacheDirectory sdirectory) {
        return sdirectory.getValueAsFile();
    }
    //////////////////////////////////////////////////////////////////////////////////////////////////




    ///////////////////////////////////////////////////////////////////////////////////////
    public static final String PRESET_PREFIX = "activity";
    public static final String PREVIEW_EXTENSION = ".preview";

    static public Foc getTrackListDirectory(SolidDataDirectory sdirectory, int i) {
        return getDataDirectory(sdirectory,getPresetPrefix(i));
    }
    static private String getPresetPrefix(int i) {
        return PRESET_PREFIX + i;
    }
    ////////////////////////////////////////////////////////////////////////////////////////




    ///////////////////////////////////////////////////////////////////////////////////////////////

    private static final int MAX_TRY=99;
    public static final String GPX_EXTENSION=".gpx";


    public static Foc generateUniqueFilePath(Foc directory, String prefix, String extension)
            throws IOException {

        Foc file = directory.child(generateFileName(prefix, extension));

        int x=1;

        while (file.exists() && x < MAX_TRY) {
            file = directory.child(generateFileName(prefix, x, extension));
            x++;
        }

        if (file.exists()) throw new IOException();

        return file;
    }


    public static String generateDatePrefix() {
        long time = System.currentTimeMillis();
        return String.format(Locale.ROOT,
                "%tY_%tm_%td_%tH_%tM", time, time, time, time, time);
    }


    private static String generateFileName(String prefix, String extension) {
        return prefix + extension;
    }

    private static String generateFileName(String prefix, int i, String extension) {
        return String.format(Locale.ROOT,
                "%s_%d%s", prefix, i, extension);
    }



    public static String parsePrefix(Foc file) {
        StringBuilder name = new StringBuilder(file.getName());

        int length=name.length();

        for (int i=length-1; i>0; i--) {
            if (name.charAt(i)=='.') {
                length=i;
                break;
            }
        }

        name.setLength(length);

        return name.toString();
    }

}

File: aat-lib/src/main/java/ch/bailu/aat_lib/util/fs/AppDirectory.kt
package ch.bailu.aat_lib.util.fs

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.preferences.map.SolidTileCacheDirectory
import ch.bailu.aat_lib.preferences.presets.SolidPreset
import ch.bailu.aat_lib.preferences.system.SolidDataDirectory
import ch.bailu.aat_lib.resources.Res
import ch.bailu.foc.Foc

object AppDirectory {
    //////////////////////////////////////////////////////////////////////////////////////
    fun getDataDirectory(sdirectory: SolidDataDirectory, sub: String): Foc {
        return sdirectory.getValueAsFile().descendant(sub)
    }

    const val DIR_AAT_DATA = "aat_data"
    const val DIR_TILES = "tiles"
    const val DIR_TILES_OSMDROID = "osmdroid/tiles"
    private const val DIR_LOG = "log"
    private const val FILE_LOG = "log.gpx"

    @JvmStatic
    fun getLogFile(sdirectory: SolidDataDirectory): Foc {
        return getDataDirectory(sdirectory, DIR_LOG).child(FILE_LOG)
    }

    const val DIR_OVERLAY = "overlay"
    const val DIR_IMPORT = "import"

    const val DIR_QUERY = "query"
    const val DIR_NOMINATIM = "${DIR_QUERY}/nominatim"
    const val FILE_NOMINATIM = "nominatim.xml"
    const val FILE_NOMINATIM_REVERSE = "reverse.json"

    const val DIR_OVERPASS = "${DIR_QUERY}/overpass"
    const val FILE_OVERPASS = "overpass.xml" // Extension .osm is not compatible with Android SAF

    const val DIR_POI = "${DIR_QUERY}/poi"
    const val FILE_POI = "poi.gpx"

    const val FILE_CRITICAL_MAP = "cm.json"
    const val FILE_BROUTER = "brouter.gpx"

    const val DIR_CACHE = "cache"
    const val FILE_CACHE_DB = "summary.db"
    const val FILE_CACHE_MVDB = "summary.mv.db"
    const val DIR_EDIT = "overlay/draft"
    const val FILE_EDIT_BACKUP = "edit.txt"
    const val FILE_DRAFT = "draft.gpx"
    const val FILE_SELECTION = "selection.txt"
    fun getEditorDraft(sdirectory: SolidDataDirectory): Foc {
        return getDataDirectory(sdirectory, DIR_EDIT).child(FILE_DRAFT)
    }

    /////////////////////////////////////////////////////////////////////////////////////////////////
    @JvmStatic
    fun getTileFile(tilePath: String?, sdirectory: SolidTileCacheDirectory): Foc {
        return getTileCacheDirectory(sdirectory).descendant(tilePath)
    }

    private fun getTileCacheDirectory(sdirectory: SolidTileCacheDirectory): Foc {
        return sdirectory.getValueAsFile()
    }

    ///////////////////////////////////////////////////////////////////////////////////////
    const val PRESET_PREFIX = "activity"
    const val PREVIEW_EXTENSION = ".preview"
    fun getTrackListDirectory(sdirectory: SolidDataDirectory, i: Int): Foc {
        return getDataDirectory(sdirectory, getPresetPrefix(i))
    }

    private fun getPresetPrefix(i: Int): String {
        return PRESET_PREFIX + i
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////
    const val GPX_EXTENSION = ".gpx"
    const val OSM_EXTENSION = ".xml" // Extension .osm is not compatible with Android SAF


    /**
     * List of gpx directories as name / directory pair
     * All presets (according to SolidPreset) and overlay and import directories
     */
    fun getGpxDirectories(appContext: AppContext): ArrayList<GpxDirectoryEntry> {
        val result = ArrayList<GpxDirectoryEntry>()
        val solidPreset = SolidPreset(appContext.storage)

        for (index in 0 until  solidPreset.length()) {
            val name = solidPreset.getValueAsString(index)
            val directory = getTrackListDirectory(appContext.dataDirectory, index)
            result.add(GpxDirectoryEntry(name, directory))
        }
        getDataDirectory(appContext.dataDirectory, DIR_OVERLAY).apply {
            result.add(GpxDirectoryEntry(Res.str().p_overlay(), this))
        }
        getDataDirectory(appContext.dataDirectory, DIR_IMPORT).apply {
            result.add(GpxDirectoryEntry(this.name, this))
        }
        return result
    }

    fun getMimeTypeFromFileName(name: String): String {
        return if (name.endsWith(GPX_EXTENSION)) {
            "application/gpx+xml" }
        else if (name.endsWith(OSM_EXTENSION)) {
            "application/xml"
        } else {
            return "application/octet-stream"
        }
    }

    data class GpxDirectoryEntry(val name: String, val file: Foc)
}
